[feature]: Guided SOQL auto-completion for Data Export

RubenHalman opened the pull request:
This Pull Request addresses Issue #912 

User is now fully guided through SOQL keywords
Includes and Excludes are displayed as suggestion when the selected field in a WHERE clause is a multi-picklist field

- [x] I have read and understand the [Contributions section](https://github.com/tprouvot/Salesforce-Inspector-reloaded#contributions)
- [x] Target branch is releaseCandidate and not master
- [x] I have performed a self-review of my code
- [x] I ran the [unit tests](https://github.com/tprouvot/Salesforce-Inspector-reloaded#unit-tests) and my PR does not break any tests
- [x] I documented the changes I've made on the [CHANGES.md](https://github.com/tprouvot/Salesforce-Inspector-reloaded/blob/master/CHANGES.md) and followed actual conventions

<img width="1258" height="404" alt="guided1" src="https://github.com/user-attachments/assets/551647bd-851b-4ed0-a3c9-f0e4f6933076" />
<img width="1258" height="404" alt="guided2" src="https://github.com/user-attachments/assets/205c9009-5120-4b38-938b-8ed0e299526b" />
<img width="1258" height="404" alt="guided3" src="https://github.com/user-attachments/assets/842027ae-7c14-45ce-a296-63e7de74d3f5" />
<img width="1258" height="404" alt="guidedincludes" src="https://github.com/user-attachments/assets/eb5d267b-588c-4507-8260-73c1fffcc1f6" />

I’m unsure about the tests: the export test is skipped, but I believe this PR does not affect the import functionality.
<img width="1248" height="703" alt="image" src="https://github.com/user-attachments/assets/1f9deaa0-e573-4d48-9528-36b7e8e8c269" />

RubenHalman committed:
[feature]: Guided SOQL auto-completion

RubenHalman commented:
I’m not sure I’m fully familiar with this project’s conventions and making updates to `changes.md`. Any guidance or feedback would be appreciated.

tprouvot commented:
Hi @RubenHalman 
Thanks for the PR !

Regarding the project conventions, you can take a look at the [contribution sections ](https://github.com/tprouvot/Salesforce-Inspector-reloaded?tab=readme-ov-file#contributions) from README.

I'll get back to you with a review during the week

RubenHalman commented on addon/data-export.js:
```diff
@@ -425,72 +425,147 @@ class Model {
+    if (beforeCursor === "" || (beforeCursor.length < 10 && !beforeCursor.includes("select") && !beforeCursor.includes("find"))) {
+      let keywords = [
+        {value: "SELECT", title: "SELECT - Start a SOQL query", suffix: " ", rank: 1, autocompleteType: "keyword", dataType: ""},
+        {value: "FIND", title: "FIND - Start a SOSL search", suffix: " ", rank: 2, autocompleteType: "keyword", dataType: ""}
```
I haven’t tested SOSL yet. I’ll double-check tomorrow.

RubenHalman commented on addon/data-export.js:
```diff
@@ -425,72 +425,147 @@ class Model {
+      // After typing some fields, suggest FROM
+      if (afterSelect.length > 0 && !afterSelect.endsWith(",") && !afterSelect.endsWith(" ")) {
+        let suggestions = [
+          {value: ", ", title: "Add another field", suffix: "", rank: 1, autocompleteType: "punctuation", dataType: ""},
```
comma is never automatically inserted when selecting suggested fields, so user may end up with `SELECT DandbCompany.City AnnualRevenue FROM Account` instead of `SELECT DandbCompany.City, AnnualRevenue FROM Account`

RubenHalman commented on addon/data-export.js:
```diff
@@ -425,72 +425,147 @@ class Model {
+          {value: "Id FROM", title: "Id FROM - Quick start (then replace Id with your fields)", suffix: " ", rank: 1, autocompleteType: "pattern", dataType: ""},
+          {value: "Id", title: "Id - Record identifier", suffix: ", ", rank: 2, autocompleteType: "fieldName", dataType: "id"},
+          {value: "Name", title: "Name - Record name", suffix: ", ", rank: 2, autocompleteType: "fieldName", dataType: "string"},
+          {value: "Id, Name FROM", title: "Id, Name FROM - Common starter", suffix: " ", rank: 2, autocompleteType: "pattern", dataType: ""},
```
The Id and Name options should not insert comma's when object remains to be selected.

tprouvot replied:
I think comma should never be automatically inserted, this was the reason why I created the fork 😄

RubenHalman replied:
Hmm upon reviewing your input and the code again, I think that for an optimized user experience, we can guide the user so that if he or she selects multiple field, the comma gets automatically added in between those, but when mouse/keyboard leaves the query we remove the comma after the last field selected.

RubenHalman committed:
revert comment removal

RubenHalman committed:
revert comment removal

RubenHalman committed:
revert comnments

RubenHalman commented on addon/data-export.js:
```diff
@@ -639,46 +729,125 @@ class Model {
         return;
       }
       if (sobjectStatuses.has("notfound")) {
-        vm.autocompleteResults = {
```
perhaps I should revert these changes to the fomatting

tprouvot commented:
Great work @RubenHalman I like the changes you developed !

tprouvot commented on addon/data-export.js:
```diff
@@ -778,111 +937,81 @@ class Model {
-          yield {value: "NEXT_N_FISCAL_YEARS:n", title: "Starts 12:00:00 on the first day of the next fiscal year and continues through the end of the last day of the nth fiscal year. The fiscal year is defined in the company profile under Setup at Company Profile | Fiscal Year.", suffix: " ", rank: 1, autocompleteType: "variable", dataType: ""};
-          yield {value: "LAST_N_FISCAL_YEARS:n", title: "Starts 12:00:00 on the first day of the last fiscal year and continues through the end of the last day of the previous nth fiscal year. The fiscal year is defined in the company profile under Setup at Company Profile | Fiscal Year.", suffix: " ", rank: 1, autocompleteType: "variable", dataType: ""};
-          yield {value: "N_FISCAL_YEARS_AGO:n", title: "Starts at 12:00:00 AM on the first day of the fiscal year n fiscal years ago and continues through the end of the last day of that fiscal year.", suffix: " ", rank: 1, autocompleteType: "variable", dataType: ""};
+          yield {value: "YESTERDAY", title: "Yesterday", suffix: " ", rank: 1, autocompleteType: "variable", dataType: ""};
```
Why did you removed many date literals ? Maybe an error ?

RubenHalman replied:
Yes, you’re right: I regret the formatting changes. I mistakenly thought this was one of those. I’ll revert those (and this) shortly. Thank you, and my apologies.

tprouvot commented:
There is a behavior change in the sobject name suggestion that make the selection less convenient, I think we should keep existing behaviour

Before :
<img width="674" height="390" alt="image" src="https://github.com/user-attachments/assets/dc6b6ff7-7e5f-4632-b495-512c6933a0fe" />


After:
<img width="878" height="339" alt="image" src="https://github.com/user-attachments/assets/324c6c0e-b771-4da7-9a5f-c4b786375f1f" />

RubenHalman committed:
comma handling

tprouvot commented:
> There is a behavior change in the sobject name suggestion that make the selection less convenient, I think we should keep existing behaviour
> 
> Before : <img alt="image" width="674" height="390" src="https://private-user-images.githubusercontent.com/35368290/531883265-dc6b6ff7-7e5f-4632-b495-512c6933a0fe.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njc2MjQ0NjIsIm5iZiI6MTc2NzYyNDE2MiwicGF0aCI6Ii8zNTM2ODI5MC81MzE4ODMyNjUtZGM2YjZmZjctN2U1Zi00NjMyLWI0OTUtNTEyYzY5MzNhMGZlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAxMDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMTA1VDE0NDI0MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc1ZWFmNzQ3ZTQ4OGQ2N2VjZjI2MjA1YWMyZmE0M2YyMzAxNWMyOGU3NjA5N2QzZDhlNTczNGZiYTJkMjVkODImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.gO1wxR47lcf3HV8OfgIfMIamOSY6SfIptfZtPb9tMpI">
> 
> After: <img alt="image" width="878" height="339" src="https://private-user-images.githubusercontent.com/35368290/531883220-324c6c0e-b771-4da7-9a5f-c4b786375f1f.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njc2MjQ0NjIsIm5iZiI6MTc2NzYyNDE2MiwicGF0aCI6Ii8zNTM2ODI5MC81MzE4ODMyMjAtMzI0YzZjMGUtYjc3MS00ZGE3LTlhNWYtYzRiNzg2Mzc1ZjFmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAxMDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMTA1VDE0NDI0MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM2ZjM2YTVjZGExMjljODQ2MmRhZTk4NWM2OGNlNjgwMjUyMDg4Yjg2NTNkZDlmMzI4MmMxNmExMjdkNTZlNTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.PYLdKLdpnLc_cWK3uB-EJSBcYlq-GIAc7DO1q55mRhk">

Not sure if my comment was clear but by "existing behaviour" I was referring to the one we have in releaseCandidate (first screenshot)

RubenHalman commented:
@tprouvot Thank you for your rapid feedback. My apologies, I should have probably put it back in draft and was a bit over optimistic on when I would be able to make all improvements, but I agree 100% and am working on it. Thank you again!

RubenHalman committed:
revert formatting and date literal changes

RubenHalman committed:
correct word placement when picking suggestions

RubenHalman committed:
insert comma when running export

RubenHalman committed:
auto normalize double spaces

RubenHalman commented:
I’ve reverted the previous behavior as requested. All together, this PR has become a bit more code-heavy than I’d like, but I think the improved experience is worth it.

Some of the changes added are:

- SOQL suggestions now appear throughout all stages, depending on the cursor position.
- Object suggestion behavior has been reverted to match the previous experience.
- Commas are never added during field selection, but any missing commas are automatically added on Run Export.
- Double spaces are automatically removed to prevent any race condition in auto-suggestions. 

Multi-line queries still execute but are reformatted after execution. I’m not sure if this is an issue. If it is, we could introduce a toggle to let users differentiate. I’d be happy to hear your thoughts on this.

RubenHalman commented on addon/data-export.js:
```diff
@@ -384,6 +384,21 @@ class Model {
     let selEnd = vm.queryInput.selectionEnd;
     let ctrlSpace = e.ctrlSpace;
 
+    // Automatically normalize double spaces
```
soql does not care about double spaces or line breaks so this should be supported in data export, without having to normalize

RubenHalman committed:
prevents issues when cursor is in whitespace

RubenHalman commented:
Hi @tprouvot,

I’ve reverted the previous behavior as requested. It’s a bit more code-heavy than I’d like, but I think the improved experience is worth it. The changes added are:

- Smart SOQL and auto complete suggestions appear throughout all stages, depending on the cursor position.
- Commas are no longer added during field selection, but any missing commas are automatically added on Run Export.
- Reverted the removal of date literals.

Thank you!

RubenHalman committed:
add changelog entry to changes.md
